homeassistant/components/google/__init__.py

```diff
@@ -213,24 +239,25 @@ def setup(hass, config):
     if not os.path.isfile(token_file):
         do_authentication(hass, config, conf)
     else:
-        if not check_correct_scopes(token_file):
+        if not check_correct_scopes(token_file, conf):
             do_authentication(hass, config, conf)
         else:
             do_setup(hass, config, conf)
 
     return True
 
 
-def check_correct_scopes(token_file):
+def check_correct_scopes(token_file, config):
     """Check for the correct scopes in file."""
     with open(token_file) as tokenfile:
-        if "readonly" in tokenfile.read():
+        contents = tokenfile.read()
+        if (config.get(CONF_CALENDAR_ACCESS) is FeatureAccess.ReadWrite and "readonly" in contents) or (config.get(CONF_CALENDAR_ACCESS) is FeatureAccess.Read and "readonly" not in contents):
```

Comment:
I would suggest making the above change then doing a check for the entire scope url to simplify this logic, and perhaps make it less error prone.  That is, I think what this is trying to do is assert that the tokenfile contains the desired scope, so explicitly checking if the desired scope is in the file may be more straight forward.

(I realize this PR extends the existing pattern, so i'm more suggesting the existing pattern is a little brittle)

Reply:
Updated to hopefully be more robust